### PR TITLE
feat(velero): velero plugin for backup and restore for ZFS-LocalPV

### PR DESCRIPTION
**Pull request by pawanpraka1**
**Why is this PR required? What issue does it fix?**:

Adding backup and restore support for ZFS-LocalPV. The PR in the ZFS-LocalPV repo is here : https://github.com/openebs/zfs-localpv/pull/162

**What this PR does?**:
A new provider(openebs.io/zfspv-blockstore) has been added to
support backup and restore for ZFS-LocalPV. We can use below
VolumeSnapshotLocation to backup/restore the data to/from the cloud.

```yaml
apiVersion: velero.io/v1
kind: VolumeSnapshotLocation
metadata:
  name: default
  namespace: velero
spec:
  provider: openebs.io/zfspv-blockstore
  config:
    bucket: velero
    prefix: zfs
    namespace: openebs # this is namespace where ZFS-LocalPV creates all the CRs, passed as OPENEBS_NAMESPACE env in the ZFS-LocalPV deployment
    provider: aws
    region: minio
    s3ForcePathStyle: "true"
    s3Url: http://minio.velero.svc:9000
```

**Does this PR require any upgrade changes?**:

no

**If the changes in this PR are manually verified, list down the scenarios covered and commands you used for testing with logs:**

- velero full backup and restore
- velero schedule(full) backup and restore 

test case: backup the full namespace, delete it, and restore it and verify the data. 

**Any additional information for your reviewer?**:

- Restore in a namespace where volume already exist, will fail. 

I am using local repo in go mod as ZFS-LocalPV side changes are not merged. Will update that once changes are merged.

github.com/openebs/zfs-localpv => /home/pawan/Desktop/openebs/src/github.com/openebs/zfs-localpv

**To Do**:
- velero schedule(incremental) backup and restore 

There are few changes related to it in this PR, will support the full incremental backup and restore after PR https://github.com/openebs/velero-plugin/pull/99

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

File: pkg/zfs/plugin/backup.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2020 the Velero contributors.
```
Comment (mynktl) [inline pkg/zfs/plugin/backup.go:2]:
```suggestion
Copyright 2020 The OpenEBS Authors.
```

File: pkg/zfs/plugin/zfs.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+	p.remoteAddr, _  = utils.GetServerAddress()
+	if p.remoteAddr == "" {
+		return errors.New("zfs: error fetching Server address")
+	}
+
+	if ns, ok := config[ZFSPV_NAMESPACE]; ok {
+		p.namespace = ns
+	} else {
+		p.namespace = "openebs" // default namespace
```
Comment (mynktl) [inline pkg/zfs/plugin/zfs.go:78]: better to use constant in place of hardcoded string

File: pkg/zfs/plugin/zfs.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+// GetVolumeInfo returns the type and IOPS (if using provisioned IOPS) for
+// the specified volume in the given availability zone.
+func (p *Plugin) GetVolumeInfo(volumeID, volumeAZ string) (string, *int64, error) {
+	p.Log.Infof("zfs: GetVolumeInfo called", volumeID, volumeAZ)
+	return "zfs-localpv", nil, nil
+}
+
+// IsVolumeReady Check if the volume is ready.
+func (p *Plugin) IsVolumeReady(volumeID, volumeAZ string) (ready bool, err error) {
+	p.Log.Infof("zfs: IsVolumeReady called", volumeID, volumeAZ)
```
Comment (mynktl) [inline pkg/zfs/plugin/zfs.go:131]: this logging can be removed or moved to debug level.

File: pkg/zfs/plugin/zfs.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+	return p.isVolumeReady(volumeID)
+}
+
+// CreateSnapshot creates a snapshot of the specified volume, and applies any provided
+// set of tags to the snapshot.
+func (p *Plugin) CreateSnapshot(volumeID, volumeAZ string, tags map[string]string) (string, error) {
+	p.Log.Infof("zfs: CreateSnapshot called", volumeID, volumeAZ, tags)
```
Comment (mynktl) [inline pkg/zfs/plugin/zfs.go:139]: better to move it to debug level.

File: pkg/zfs/plugin/restore.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	restoreStatusInterval = 5
+)
+
+func (p *Plugin) createVolume(pvname string, bkpname string, bkpZV *apis.ZFSVolume) (*apis.ZFSVolume, bool, error) {
+
+	// get the target namespace
+	ns, err := velero.GetRestoreNamespace(bkpZV.Labels[VeleroNsKey], bkpname, p.Log)
+
+	if err != nil {
+		p.Log.Errorf("zfs: failed to get target ns pv %s bkpname %s err: %v", pvname, bkpname, err)
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:42]:
```suggestion
		p.Log.Errorf("zfs: failed to get target ns for pv=%s, bkpname=%s err: %v", pvname, bkpname, err)
```

File: pkg/zfs/plugin/restore.go [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+	if vol == nil {
+		// this is first full restore, go ahead and create the volume
+		rZV := &apis.ZFSVolume{}
+		// hack(https://github.com/vmware-tanzu/velero/pull/2835): generate a new uuid only if PV exist
+		pv, err := p.getPV(pvname)
+
+		if err == nil && pv != nil {
+			rvol, err := utils.GetRestorePVName()
+			if err != nil {
+				return nil, false, errors.Errorf("zfs: failed to get restore vol name for %s", pvname)
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:80]:
```suggestion
				return nil, false, errors.Errorf("zfs: failed to generate restore vol name for %s", pvname)
```

File: pkg/zfs/utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	backupName = s[1]
+
+	if volumeID == "" || backupName == "" {
+		err = errors.Errorf("invalid volumeID=%s backupName=%s", volumeID, backupName)
+	}
+	return
+}
+
+// GetRestorePVName return new name for clone pv for the given pv
+func GetRestorePVName() (string, error) {
```
Comment (mynktl) [inline pkg/zfs/utils/utils.go:75]: better to change this function name, like generatePvName..

File: pkg/zfs/plugin/backup.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	go p.checkBackupStatus(bkpname)
+
+	p.Log.Infof("zfs: uploading Snapshot %s file %s", snapname, filename)
+	ok := p.cl.Upload(filename, size)
+	if !ok {
+		p.deleteBackup(bkpname)
```
Comment (mynktl) [inline pkg/zfs/plugin/backup.go:233]: need to log error of deleteBackup for debugging purpose or manual cleanup.

File: pkg/zfs/plugin/backup.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+		}
+		size := len(backups)
+		sort.Strings(backups)
+		return backups[size - 1], nil
+	}
+
+	return "", nil
+}
+
+func (p *Plugin) createBackup(vol *apis.ZFSVolume, schdname string, snapname string) (string, error) {
```
Comment (mynktl) [inline pkg/zfs/plugin/backup.go:104]: nit:
```suggestion
func (p *Plugin) createBackup(vol *apis.ZFSVolume, schdname, snapname string) (string, error) {
```

File: pkg/zfs/plugin/zfs.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	"github.com/openebs/velero-plugin/pkg/zfs/utils"
+	"github.com/openebs/velero-plugin/pkg/velero"
+	cloud "github.com/openebs/velero-plugin/pkg/clouduploader"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// ZFSPV_NAMESPACE config key for OpenEBS namespace
+	ZFSPV_NAMESPACE = "namespace"
```
Comment (mynktl) [inline pkg/zfs/plugin/zfs.go:34]: nit: 
```suggestion
	ZfsPvNamespace = "namespace"
```

File: pkg/zfs/plugin/zfs.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+// set of tags to the snapshot.
+func (p *Plugin) CreateSnapshot(volumeID, volumeAZ string, tags map[string]string) (string, error) {
+	p.Log.Infof("zfs: CreateSnapshot called", volumeID, volumeAZ, tags)
+
+	bkpname, ok := tags[VeleroBkpKey]
+	if !ok {
+		return "", errors.New("zfs: error get backup name")
+	}
+
+	schdname, ok := tags[VeleroSchdKey]
```
Comment (mynktl) [inline pkg/zfs/plugin/zfs.go:146]:
```suggestion
	schdname, _ := tags[VeleroSchdKey]
```

File: pkg/zfs/plugin/zfs.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+	if snapshotID == "" {
+		p.Log.Warning("zfs: Empty snapshotID")
+		return nil
+	}
+
+	return p.deleteBackup(snapshotID)
+}
+
+// GetVolumeID returns the specific identifier for the PersistentVolume.
+func (p *Plugin) GetVolumeID(unstructuredPV runtime.Unstructured) (string, error) {
```
Comment (mynktl) [inline pkg/zfs/plugin/zfs.go:171]: need to check if PV is provisioned by zfs-localpv driver or not. Refer: https://github.com/openebs/velero-plugin/blob/master/pkg/cstor/cstor.go#L275

File: pkg/zfs/plugin/backup.go (hunk 8 above)
Comment (pawanpraka1) [reply pkg/zfs/plugin/backup.go:233]: already loggin it inside the funtion.

File: pkg/zfs/plugin/backup.go [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+		size := len(backups)
+		sort.Strings(backups)
+		return backups[size - 1], nil
+	}
+
+	return "", nil
+}
+
+func (p *Plugin) createBackup(vol *apis.ZFSVolume, schdname, snapname string) (string, error) {
+
```
Comment (mynktl) [inline pkg/zfs/plugin/backup.go:105]: nit: unnecessary leading newline

File: pkg/zfs/plugin/restore.go [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+	"github.com/openebs/zfs-localpv/pkg/builder/restorebuilder"
+	apis "github.com/openebs/zfs-localpv/pkg/apis/openebs.io/zfs/v1"
+)
+
+const (
+	restoreStatusInterval = 5
+)
+
+func (p *Plugin) createVolume(pvname string, bkpname string, bkpZV *apis.ZFSVolume) (*apis.ZFSVolume, error) {
+
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:37]: nit: unnecessary new line

File: pkg/zfs/plugin/restore.go [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	return vol.Status.State == zfs.ZFSStatusReady, nil
+}
+
+func (p *Plugin) checkRestoreStatus(snapname string) {
+
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:154]: nit: unnecessary new line

File: pkg/zfs/plugin/restore.go [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+			p.Log.Errorf("zfs: update ZFSVolume failed vol %v err: %v", vol, err)
+			return nil, err
+		}
+	}
+
+	return vol, nil
+}
+
+func (p *Plugin) restoreZFSVolume(pvname, bkpname string) (*apis.ZFSVolume, error) {
+
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:124]: nit: unnecessary new line

File: pkg/zfs/plugin/restore.go [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	return vol.Status.State == zfs.ZFSStatusReady, nil
+}
+
+func (p *Plugin) checkRestoreStatus(snapname string) {
+
+	for true {
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:155]:
```suggestion
	for {
```

File: pkg/zfs/plugin/restore.go [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+
+	if err != nil {
+		p.Log.Errorf("zfs: failed to get source volume failed vol %s snap %s err: %v", pvname, bkpname, err)
+		return nil, err
+	}
+
+	var vol *apis.ZFSVolume = nil
+
+	if len (volList.Items) > 1 {
+		return nil, errors.Errorf("zfs: error can not have more than one source volume %s", pvname, bkpname)
```
Comment (mynktl) [inline pkg/zfs/plugin/restore.go:59]: missing indentifier for `bkpname` in `Errorf`.

**Review by mynktl — APPROVED** [review]
looks good. Given few comments about formatting.

**Review by mynktl — APPROVED** [review]

File: Makefile [hunk 19]
```diff
@@ -83,7 +83,7 @@ all: build
 container: all
 	@echo ">> building container"
 	@cp Dockerfile _output/Dockerfile
-	docker build -t $(IMAGE):$(IMAGE_TAG) ${DBUILD_ARGS} -f _output/Dockerfile _output
+	@sudo docker build -t $(IMAGE):$(IMAGE_TAG) ${DBUILD_ARGS} -f _output/Dockerfile _output
```
Comment (kmova) [inline Makefile:86]: Are these changes really required? sudo?

**Review by kmova — APPROVED** [review]

File: Makefile (hunk 19 above)
Comment (pawanpraka1) [reply Makefile:86]: yes, needed for docker.
